src/v2/blobs.rs

```diff
@@ -62,60 +47,32 @@ impl Client {
-        let fres = freq
-            .from_err()
-            .inspect(move |_| {
-                trace!("GET {:?}", url);
-            }).and_then(move |r| {
-                match r.status() {
-                    StatusCode::MOVED_PERMANENTLY
-                    | StatusCode::TEMPORARY_REDIRECT
-                    | StatusCode::FOUND => {
-                        trace!("Got moved status {:?}", r.status());
-                    }
-                    _ => return Either::A(future::ok(r)),
-                };
-                let redirect: Option<String> = match r.headers().get("Location") {
-                    None => {
-                        return Either::A(future::err(Error::from(
-                            "get_blob: missing location header",
-                        )))
-                    }
-                    Some(loc) => {
-                        trace!("Got Location header {:?}", loc);
-                        String::from_utf8(loc.as_bytes().to_vec()).ok()
+
+        let fres = reqwest::async::Client::new()
+            .get(url)
+            .send()
+            .inspect(|res| trace!("Blob GET status: {:?}", res.status()))
+            .and_then(|mut res| {
+                let body =
+                    std::mem::replace(res.body_mut(), reqwest::async::Decoder::empty()).concat2();
+                body.join(futures::future::ok(res))
+            }).map_err(|e| Error::from(format!("{}", e)))
+            .and_then(|(body, res)| {
+                let body_vec = body.to_vec();
+                let len = body_vec.len();
+                match res.status() {
+                    StatusCode::OK => {
+                        trace!("Received {} bytes from blob", len);
+                        Ok(body_vec)
                     }
-                };
-                if let Some(u) = redirect {
-                    let new_url = match hyper::Uri::from_str(u.as_str()) {
-                        Ok(u) => u,
-                        _ => {
-                            return Either::A(future::err(
-                                format!("get_blob: wrong URL '{}'", u).into(),
-                            ))
-                        }
-                    };
-                    trace!("Following redirection to {}", new_url);
-                    let mut req = hyper::Request::default();
-                    *req.method_mut() = hyper::Method::GET;
-                    *req.uri_mut() = new_url;
-                    return Either::B(cl.hclient.request(req).from_err());
-                };
-                Either::A(future::ok(r))
-            }).and_then(|r| {
-                r.into_body()
-                    .concat2()
-                    .map_err(|e| format!("get_blob: failed to fetch the whole body: {}", e).into())
-            }).and_then(|body| Ok(body.into_bytes().to_vec()));
+                    status => Err(Error::from(format!(
+                        "GET request failed with status '{}' and body of size {}: {:#?}",
```

Comment:
Minor nit: I think that this kind of status check is usually performed before starting to collect the whole body (to skip useless work) and the return error doesn't usually carry around the whole response (which could potentially be quite huge). I don't know if this can apply here and I'll leave the ergonomics to you, I just wanted to mention that.

Reply:
https://www.freesoft.org/CIE/RFC/2068/89.htm contains that on 4xx the body should be displayed to the user. I'll give filtering this a shot.

Reply:
Ack. I'm fine both ways. Leaving this to your discretion.